### PR TITLE
[mongodb] loosen PullOperator (fixes nested arrays)

### LINKED ISSUE #29911: MongoDB $pull won't work for nested interface types.
If you know how to fix the issue, make a pull request instead.

- [x] I tried using the `@types/mongodb` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @CaselIT @daprahamian @andy-ms @bitjson @dante-101 @mcortesi @EnricoPicci @AJCStriker @julien-c @daprahamian @denys-bushulyak @BastienAr @sindbach @geraldinelemeur @jishi @various89 @angela-1 @lirbank @hector7 @floric

If you do not mention the authors the issue will be ignored.

```typescript
interface BundleDocument {
  _id: ObjectId;
  items: BundleItemDocument[];
}

interface BundleItemDocument {
  product: string;
  sku: string;
  price: number;
}

const collection = db.collection<BundleDocument>('bundles');

await collection.updateOne({ _id }, {
  $pull: {
    items: {
      product: "..."
    }
  }
});
```

The above shows incompabitle index signatures.  I believe it's when items are nested?

I'm trying to remove an "item" where "product" = "..." and @types/mongodb won't let me.

### PR DESCRIPTION
**Pull request by j**
This fixes: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/29911

I just made it looser, and you still get typings.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/29911

Authors: @CaselIT @daprahamian @andy-ms @bitjson @Dante-101 @mcortesi @EnricoPicci @AJCStriker @julien-c @daprahamian @Denys-Bushulyak @BastienAr @sindbach @geraldinelemeur @jishi @various89 @angela-1 @lirbank @hector7 @floric

### CODE REVIEW

File: types/mongodb/index.d.ts [hunk 1]
```diff
@@ -1204,7 +1204,7 @@ export type PushOperator<TSchema> = ({
 export type PullOperator<TSchema> = ({
     readonly [key in KeysOfAType<TSchema, any[]>]?: Unpacked<TSchema[key]> | QuerySelector<Unpacked<TSchema[key]>>;
 } &
-    NotAcceptedFields<TSchema, any[]>) & {
+    NotAcceptedFields<TSchema, any[]>) | {
```
Comment (HosseinAgha) [inline types/mongodb/index.d.ts]: I think you have just opened the types to accept anything and reverted the previous typings.
Please add some failing tests to updateX files. e.g. you should not accept string in a number field. Checkout ‘filterQuery’ file on how to write failing tests.

**Review by HosseinAgha — CHANGES_REQUESTED** [review]

File: types/mongodb/index.d.ts (hunk 1 above)
Comment (j) [reply types/mongodb/index.d.ts]: I see.  The tests I added are failing tests if you revert this simple commit.  I loosened them to at least have support for them.  Some support is better than no support?  I'd be more pro loosen then figure out how to actually make them strict and work while allowing all tests to pass.  Currently I have to @ts-ignore everything or cast stuff as "any".

**Comment by CaselIT** [comment]
I've added a couple of comments to the pr, that fix the new tests without opening the types.

@j do you have other examples that are not working so we try to also fix them?

**Comment by j** [comment]
@CaselIT That works like a charm! I added that and force pushed back to the branch.  @HosseinAgha thoughts?

File: types/mongodb/index.d.ts (hunk 1 above)
Comment (j) [reply types/mongodb/index.d.ts]: @CaselIT I was actually pretty much on this same track earlier but didn't want to spend too much time.

**Comment by HosseinAgha** [comment]
@j I still think we need that negative tests for `updateQuery` and `updateX` as we may not always catch if someone opened up some types. Writing them is as easy as this:  
```
// $ExpectError
buildUpdateQuery({ $pull: { subInterfaceArray: 12  }});
``` 
I'm OK with your PR though.

**Review by HosseinAgha — APPROVED** [review]

**Comment by CaselIT** [comment]
I agree. It could be added by in new pr focused on that.

**Review by CaselIT — APPROVED** [review]

**Pull request merged**